### PR TITLE
[Dependency Management Phase 2]: Github Actions/Workflows RFC

### PR DESCRIPTION
**Pull request by sophiewigmore**
## Summary

[Readable](https://github.com/paketo-buildpacks/rfcs/blob/dependency-management-step-two/text/dependencies/rfcs/0000-dependency-management-phase-two.md)

Outlines the new process for discovering and updating dependencies.
- Related to https://github.com/paketo-community/explorations/issues/8

## Checklist

* [X] I have viewed, signed, and submitted the Contributor License Agreement.
* [X] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [X] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [X] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (robdimsdale) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:124]: Is the logic to detect if we need to compile or not going to be in github actions configuration or in an executable in the buildpack?

Phrased another way - if you were running the executable locally would you expect the user to know whether to compile (i.e. the logic lives in the action config) or the executable will decide for them (i.e. the logic is in the executable in the buildpack)

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (robdimsdale) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:126]: This is only in the compiled/processed case, yes? Can we clarify this point in the RFC?

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (robdimsdale) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:130]: And this PR will run the usual battery of checks - integration tests etc?

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (robdimsdale) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:137]: does "multiple variants of the same dependency" refer to different architectures (i.e. stacks)? Or something else?

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (robdimsdale) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:139]: Yes, this should be simple and we should document it. I think that this will need to be run manually if we're in a rush - e.g. trying to ship a high-priority fix.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:124]: It'll be in the buildpack code, outlined in Phase 1 RFC. A user does not need to know this information, I would expect the buildpack code to contain this logic

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:130]: Good idea, I'll add.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:30]: Great question. I think this is something we need to do more exploration on.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:69]: I like this idea. @joshuatcasey and I are investigating what this looks like now and will update the RFC if we find it viable

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md]: Done in latest update.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:69]: We may need to also pass in the OS/arch as arguments to the `make` command, because we've encountered a bug in which compilation code for bundler works differently for bionic vs jammy.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (joshuatcasey) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:33]: Buildpack authors will need to specify the desired operating systems somehow. This could be specified in a file such as `<buildpack>/dependency/operating_systems.json` that looks like this:

```json
["ubuntu-18.04", "ubuntu-22.04", "macos-latest"]
```

Note that it's currently written on a single line (see https://github.com/joshuatcasey/bundler/runs/6802820650?check_suite_focus=true) but that should be fixable.

Comment (joshuatcasey) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:33]: Do we need to give some thought to those OS/arch combinations that are not available natively via github actions virtual environments (https://github.com/actions/virtual-environments/)?

UBI is built on RHEL, are we sure that binaries built on Ubuntu will work for them? https://www.redhat.com/en/blog/introducing-red-hat-universal-base-image

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (joshuatcasey) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:69]: We may also consider a `prepare` step that could do such things as install additional tools. It could be invoked as something like `make prepare os=<whatever>` so that it can handle the different operating systems.

Comment (joshuatcasey) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:69]: We may want to add a file such as `dependency/NAME` that contains the ID of the dependency.

Also, are we aware of any buildpacks that compile more than one ID at a time?

Comment (joshuatcasey) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:69]: We may also need a `prepare-pull-request` step that combines 6, 7, and 8

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (fg-j) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:115]: I'd find it easier to understand the explanation of workflows if they were pseudo-coded as actual workflow yaml, maybe with explanatory comments sprinkled throughout.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (ForestEckhardt) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:167]: I am a little curious why we would want to write a new script instead of just modifying the functionality of `jam update-dependencies` to do whatever that script would do, or even potentially adding some new command if you want to keep the previous generations functionality.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:33]: Yea, this is something I had looked into, I think we might be able to do something with the base image used in the Github Action Dockerfile

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:115]: I think that explaining the workflows this way is easier for someone less familiar 
with Github Actions details to get the point across of what the automation is responsible for, which is all I wanted for this section. This avoids shoehorning the details of the workflows, which may change when actually implemented. I think I'd prefer to leave the explanation at this level for the RFC, especially since the workflows will be implemented once and rolled out to buildpacks, rather than buildpacks implementing their own.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (fg-j) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:30]: Running compilation in containers solves part of this problem. For containers that are compatible with the operating system you're running locally, you can simply spin up the same container used in CI, even if container is based on a different OS. For example. to compile  for alpine amd64  in an alpine-based container , simply pull the compilation image onto your mac with amd64 architecture and run it. 

Things get a bit more complicated compiling for an architecture that's different from your local machine. Is it possible at all to compile for ARM64 architectures on an amd64 box? What about compiling for Windows on a non-Windows box? Devs may need to use a remote runner for these types of dependency builds, even when they're building outside the context of the CI pipeline.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (ekcasey) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:87]: If we are using dependencies directly from upstream do we need to test them? In most cases, where we aren't compiling it seems like the workflow should be as simple as generating metadata and PRing `buildpack.toml`. If we need to test anything we could run those tests as a gate on the PR.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (ekcasey) [inline text/dependencies/rfcs/0000-dependency-management-phase-two.md:149]: Have we considered making this a separate workflow that can be used in the rare cases where it is needed? Then the update/metadata workflow could be the same in all cases, the only difference being it looks for our compiled dependency instead of an upstream artifact.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (ForestEckhardt) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:167]: I guess I just think it would be better to keep as many buildpack development tool in `jam` as possible to hopefully help spread this way of doing dependencies but I am not tied to the idea.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:33]: addressed in the latest version of the rfc

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md]: A lot of the functionality of `jam update-dependencies` is spread out into smaller, composable pieces so that they can be plugged in with other functionality in order to support a bunch of different use cases.

I'd say that the "Assemble" step (see step 9 in [the overview](https://github.com/paketo-buildpacks/rfcs/blob/dependency-management-step-two/text/dependencies/rfcs/0000-dependency-management-phase-two.md#new-plan-overview)) might be the best candidate for pushing into `jam update-dependencies` instead of a Github action. It would be a command that just knows how to update the `buildpack.toml` given metadata. I'm not particularly married to it being an action or a `jam` command really

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md]: I'm not sure I understand why this should be a separate workflow step. I think it can pretty easily just be coupled with the compilation job in the workflow, such that if compilation/artifact upload step runs, then the SHA256 and URI are just updated in the metadata. It seems pretty coupled, I can't wrap my head around why we'd need a separate mechanism to do this.

The other steps such as original metadata generation, and the `buildpack.toml` update will be the same in every case

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:30]: I think we've addressed this somewhat in the RFC. Running ARM64 compilation will require a bit more work but I think is achievable with the same structure as outlined more or less.

Comment (ryanmoran) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:30]: Multi-arch builds are achievable on a single machine. This is what `jam create-stack` is leveraging. Check out the docs [here](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images) and [here](https://docs.docker.com/desktop/multi-arch/). Basically, Docker can leverage this thing called `binfmt_misc` to handle different architectures than the host. You sacrifice a bit of performance, but it should generally work.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md]: Let's leave this up to maintainers to decide.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (fg-j) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md:30]: Maybe we can take advantage of multi-arch support via that `--platform` flag to compile dependencies for ARM64 in containers without needing to self-host ARM64 Github runners.

File: text/dependencies/rfcs/0000-dependency-management-phase-two.md
Comment (fg-j) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md]: I’d say, better to be an action if that step should run in a container, better to put in `jam` if it’ll “just work” on any CLI users local workstation, given the appropriate inputs.

Comment (sophiewigmore) [reply text/dependencies/rfcs/0000-dependency-management-phase-two.md]: alright seeing as two tooling maintainers are both on board with the metadata update step as a part of `jam`, I have updated the RFC to reflect that. It does not need to run inside of a container specifically, so it probably doesn't need to be an action.

**Comment by sophiewigmore** [comment]
@dmikusa-pivotal [the latest commit](https://github.com/paketo-buildpacks/rfcs/pull/206/commits/bfadcf3c06349fedbe863fabf6294512fa01fac5) adds some generalization around where dependency data is stored so that we aren't locked into the `buildpack.toml`. Let me know if these changes seem reasonable / if this RFC is compatible with your other proposal.

**Review by ryanmoran — APPROVED** [review]

**Comment by sophiewigmore** [comment]
@paketo-buildpacks/steering-committee @paketo-buildpacks/dependencies @robdimsdale @ForestEckhardt @fg-j anything needed here? Looking for comments or approvals!

**Review by joshuatcasey — APPROVED** [review]

**Review by ForestEckhardt — APPROVED** [review]

**Review by fg-j — APPROVED** [review]

**Review by dmikusa-pivotal — APPROVED** [review]

**Review by brayanhenao — APPROVED** [review]

**Pull request merged**
